---
title: "Analysis of mouse bone marrow single-cell RNA-seq data"
author: "Edroaldo Lummertz da Rocha"
date: "February 18, 2018"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Mouse bone marrow single-cell RNA-seq

This tutorial describes how to apply CellRouter to a mouse bone marrow single-cell RNA-sequencing dataset generated by Grun et al (Cell Stem Cell 2016). The analysis reported here reproduce the trajectory analysis presented in the Figures 2 and 3 of our manuscript (Lummertz da Rocha et al, Nature Communications 2018)

```{r load, echo=FALSE, message=FALSE, warning=FALSE}
setwd("~/Documents/Projects/CellRouter/Manuscript/Submission_1/scripts/stemid/")

```

## Load processed data

CellRouter assumes that your are providing a gene expression matrix with genes as rows and single-cells as columns. It also assumes that the data is already normalized and confounding sources of variation were properly removed.

```{r load1, message=FALSE, warning=FALSE}
source('../CellRouter_Class.R')

matrix <- get(load('results/matrix.R'))
ndata <- get(load('results/normalized_expression.R'))
fdata <- get(load('results/filtered_expression.R'))
colnames(matrix) <- c('tSNE1', 'tSNE2')

```

## Preprocessing 

```{r preprocess, message=FALSE, warning=FALSE}
## removing chromossome names from gene names and averaging expression
genes <- sapply(strsplit(rownames(ndata), split='__', fixed=TRUE), function(x){x[1]})
map <- data.frame(id=rownames(ndata), symbol=genes, stringsAsFactors = FALSE)
ndata <- averageIds(ndata, map, 'symbol')

#Remove genes with zero variance across all cells
var <- apply(ndata, 1, var)
var <- var[which(var > 0)]
ndata <- ndata[names(var),]

### selecting genes to use as regulated along developmental trajectories
pca <- prcomp(t(ndata), scale=TRUE, center=TRUE)
loadings <- pca$rotation
num_pc <- 5
quantile <- 0.975
genes2use <- unique(as.vector(unlist(apply(loadings[,1:num_pc], 2, function(x){names(x[which(abs(x) >= quantile(x, quantile))])}))))
```

## Create a CellRouter object

```{r findsubpopulations, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}
#ggrn <- get(load('results/GRN.R'))
### Subpopulation identification and gene signatures with CellRouter
cellrouter <- CellRouter(expdata=ndata, annotations=colnames(ndata))
cellrouter@rdimension <- matrix
cellrouter <- findsubpopulations(cellrouter, 5, 'jaccard', 'results/kNN_network.gml')
cellrouter <- diffexpr(cellrouter, column='population', pvalue = 0.05)
markers <- findmarkers(cellrouter)
plotReducedDimension(cellrouter, 3.5, 3.5, filename='results/tSNE.pdf')

```

## Starting trajectory analysis

We selected subpopulation 20 as the starting point for trajectory identification

```{r trajectory, eval=TRUE, fig.height=5, fig.width=5, message=FALSE, warning=FALSE}

######## Trajectory Detection using CellRouter ###
libdir <- '~/Documents/Projects/CellRouter/Manuscript/Submission_1/scripts/CellRouter/'
cellrouter <- createKNN(cellrouter, 10, 'jaccard', 'results/paths/kNN_network_trajectory.gml')
filename <- "results/paths/cell_edge_weighted_network.txt"
write.table(cellrouter@graph$edges, file=filename, sep='\t', row.names=FALSE, col.names = FALSE, quote=FALSE) #input network

##select starting subpopulation
sources <- c('SP_20')
##all other subpopulations are targets
targets <- setdiff(as.vector(cellrouter@sampTab$population), sources)
methods <- c("euclidean", "maximum", "manhattan","canberra","binary", 'graph') #graph for distances in KNN
cellrouter <- findpaths(cellrouter, libdir, paste(getwd(), 'results/paths', sep='/'), method="graph")
save(cellrouter, file='results/CellRouter_StemID.R')
```

## Preprocessing of trajectories

Identification of genes and gene clusters dynamically regulated during HSC differentiation

```{r processtrajectory, eval=FALSE}
#load
cellrouter <- get(load('results/CellRouter_StemID.R'))
ranks <- c('path_cost', 'path_flow', 'rank', 'length')
cellrouter <- processtrajectories(cellrouter, genes2use, path.rank=ranks[3], 
                                  num.cells = 3, neighs = 1)
names <- unique(names(cellrouter@pathsinfo$distr))
clusters.show <- names
cellrouter <- correlationpseudotime(cellrouter, type='spearman')
cellrouter <- topgenes(cellrouter, 0.85, 0.15)
cellrouter <- smoothdynamics(cellrouter, names)
cellrouter <- clusterGenesPseudotime(cellrouter, 5)
save(cellrouter, file='results/CellRouter_StemID_Processed.R')

```

## Reconstruction of Gene Regulatory Networks
```{r GRN, eval=TRUE}
# Gene regulatory network reconstruction using a correlation-based version of CLR
# published in our previous work with CellNet (Cahan et al, Cell 2014)
ggrn <- buildGRN('Mm', ndata, genes2use, 5, 'results/GRN.R')

```


## Downstream analysis

```{r downstream0, eval=TRUE, fig.height=2, fig.width=5, message=FALSE, warning=FALSE}
## Loading cellrouter object with processed trajectories
cellrouter <- get(load('results/CellRouter_StemID_Processed.R'))

###positive and negative controls
genelist <- c('Klf1', 'Gata1', 'Gfi1b', 'Zfpm1',
              'Cebpe', 'Gfi1', 'Cebpa', 'Sfpi1', 'Mmp8')
paths <- c('SP_20.SP_10', 'SP_20.SP_18')
plotPathHeatmap2(cellrouter, paths, genelist, TRUE, 2, 4, 2, 'results/')

```

```{r downstream, message=FALSE, warning=FALSE, eval=TRUE}

## GRN score for selected transitions
tfs <- find_tfs(species = 'Mm')
transitions <- c('SP_20.SP_17','SP_20.SP_18','SP_20.SP_19', 'SP_20.SP_10')
x <- grnscores(cellrouter, tfs, transitions, direction='up', q.up=14, dir.targets='up', 
               columns=2, width=8, height=5, flip=FALSE, filename='results/lineage_regulators_score_up')
```

```{r downstream0.1, eval=TRUE, fig.height=3.5, fig.width=7, message=FALSE, warning=FALSE}
p <- 'SP_20.SP_18'
scores <- x[[p]]$scores
m2 <- plottr(cellrouter, p, x[[p]]$scores, cluster=TRUE, 2, 2.5, 10, paste('results/', p, 'up_diff_dynamics.pdf',sep=''))

```

```{r downstream2, eval=TRUE, fig.height=2, fig.width=4, message=FALSE, warning=FALSE}
paths <- c('SP_20.SP_18')
plottrajectories(cellrouter, paths, names(scores)[1:5], rescale = TRUE, columns=1, width=5, height=2, filename='results/dynamics_curve.pdf')
plottrajectories(cellrouter, paths, c('Mmp8', 'Ngp', 'Elane', 'Retnlg'), rescale = TRUE, columns=1, width=5, height=2, filename='results/dynamics_markers.pdf')
```

```{r downstream002, eval=TRUE, message=FALSE, warning=FALSE}

plotDRExpression2(cellrouter, c('Mmp8', 'Ngp', 'Elane', 'Retnlg'), TRUE, 2, 6, 4, 'results/Neurophil_StemID.pdf')
```
```{r downstream3, eval=TRUE, fig.height=2, fig.width=7, message=FALSE, warning=FALSE}
plotbranch(cellrouter, 'up','SP_20.SP_10', 'SP_20.SP_18', 2, width=5, height=4, filename='results/SP_20.SP_10_branch_dynamics.pdf')
```

```{r downstream3.1, eval=TRUE, fig.height=2, fig.width=7, message=FALSE, warning=FALSE}
plotbranch(cellrouter, 'up','SP_20.SP_18', 'SP_20.SP_10', 2, width=3, height=4, filename='results/SP_20.SP_18_branch_dynamics.pdf')
```

```{r downstream4, eval=TRUE}
### Mxd1 targets
plotDRExpression2(cellrouter, c('Mmp8', 'Mmp9', 'Cd52', 'Retnlg', 'Mxd1'), TRUE, 2, 6, 5, 'results/Mxd1_targets.pdf')

```
```{r downstream4.02, eval=TRUE, fig.height=2, fig.width=4, message=FALSE, warning=FALSE}

```

```{r downstream4.1, eval=TRUE, fig.height=3.5, fig.width=7, message=FALSE, warning=FALSE}
##Erythroblast branch (Klf1-high subpopulation)
p <- 'SP_20.SP_10'
m2 <- plottr(cellrouter, p, x[[p]]$scores, cluster=TRUE, 2, 2.5, 10, paste('results/', p, 'up_diff_dynamics.pdf',sep=''))
```

```{r downstream4.2, eval=TRUE, fig.height=2, fig.width=4, message=FALSE, warning=FALSE}
genelist <- c('Klf1', 'Gata1', 'Gata2', 'Rnf10', 'Snca')
plottrajectories(cellrouter, p, genelist, rescale = TRUE, columns=1, width=4, height=2, filename='results/dynamics_curve2.pdf')

```



```{r pathway, eval=TRUE}
#### Pathway enrichment analysis on selected trajectories
#ids and cc are provided in the github webpage
ids <- get(load('ids.R'))
cc <- get(load('cell_cycle_genes_December_30_2015.R')) #cell cycle genes are removed
paths <- c('SP_20.SP_10', 'SP_20.SP_19', 'SP_20.SP_17', 'SP_20.SP_18')
cellrouter <- pathwayenrichment(cellrouter, paths, cc, 'mouse','org.Mm.eg.db', ids)
enr <- pathwaycluster(cellrouter, cellrouter@pathwayenrichment$UP$GOBP, 10, TRUE, 5, 5, 'results/Supplementary_Table_2_GO.pdf')
```

